### PR TITLE
[Feature] Preference model e2e

### PR DESCRIPTION
**Pull request by ppppqp**
End-to-end support preference model in vLLM-SR

FIX https://github.com/vllm-project/semantic-router/issues/996

Demo:

https://github.com/user-attachments/assets/f1215f38-5e47-4b72-b3bf-4eb7a684edb3

This PR:
- Introduce the local preference model config under the `classifier` route
- Implemented multi-token zero shot classifier for the model, which is needed because the lable in preference model can be free-form, so likely to span across multiple tokens. Current `classify_zero_shot` only looks at the first token of each label.
- Uses Qwen/Qwen3-1.7B for now, which is good enough, from my own experience.

Not included:
- Model training code, since we are simply using Qwen's base model at the moment.

## Performance eval
Add several benchmark cases on inference speed:
- Short query-short label: a short input sentence with 4 preference labels.
- Short query-long label: a short input sentence with 20 preference labels.
- Very long query: a very long input sentence (1200 words) with 4 preference labels.

Baseline:
Category classifier with Qwen3-0.6B
- Short query: a short input sentence
- Very long query: a very long input sentence

To run the benchmark:
```bash
# in candle-binding/
cargo build -p candle-semantic-router --release --features cuda # if on cpu, use --no-default-features
export LD_LIBRARY_PATH=$(pwd)/target/release:$LD_LIBRARY_PATH

# in perf/benchmarks
go test -bench=BenchmarkPreference_ShortLabels -run=^$
```

Result:

On CPU (Apple M4)
test case | preference classifier | baseline classifier | rate
 |---|---|---|---|
short query|1.64s|0.53s | 3.09x
short query/long label|5.72s| / | /
long query|15.23s|7.5s | 2.03x

On GPU (Nvidia A40)
test case | preference classifier | baseline classifier | rate
 |---|---|---|---|
short query|0.068s|0.019s | 3.57x
short query/long label|0.23s| / | /
long query|0.20s|0.12s | 1.66x

Raw output from benchmark
CPU:
```plaintext
# preference classifier
BenchmarkPreference_ShortLabels-10             1        1645658792 ns/op             152 B/op          3 allocs/op
BenchmarkPreference_LongLabels-10              1        5723467209 ns/op             472 B/op          3 allocs/op
BenchmarkPreference_VeryLongQuery-10                   1        15238879416 ns/op            152 B/op        3 allocs/op


# baseline classifier
BenchmarkCategoryAdapter_Short-10              2         534199958 ns/op             192 B/op          4 allocs/op
BenchmarkCategoryAdapter_VeryLongQuery-10              1        7513401833 ns/op             192 B/op        4 allocs/op
```
GPU:
```plaintext
# preference classifier
BenchmarkPreference_ShortLabels-96            16          68354983 ns/op             163 B/op          3 allocs/op
BenchmarkPreference_LongLabels-96              5         238900429 ns/op             590 B/op          4 allocs/op
BenchmarkPreference_VeryLongQuery-96                   5         206720263 ns/op             152 B/op          3 allocs/op

# baseline classifier
BenchmarkCategoryAdapter_Short-96             57          19898056 ns/op             192 B/op          4 allocs/op
BenchmarkCategoryAdapter_VeryLongQuery-96              8         125659917 ns/op             192 B/op          4 allocs/op
```

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

- [ ] Make sure the code changes pass the [pre-commit](https://github.com/vllm-project/semantic-router/blob/main/CONTRIBUTING.md) checks.
- [ ] Sign-off your commit by using <code>-s</code> when doing <code>git commit</code>
- [ ] Try to classify PRs for easy understanding of the type of changes, such as `[Bugfix]`, `[Feat]`, and `[CI]`.

<details>

<summary><b> Detailed Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to semantic-router! Before submitting the pull request, please ensure the PR meets the following criteria. This helps us maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Please try to classify PRs for easy understanding of the type of changes. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[CLI]</code> for changes to the command-line interface tools.</li>
    <li><code>[Dashboard]</code> for changes to the dashboard or web UI.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Feat]</code> for new features in the cluster (e.g., autoscaling, disaggregated prefill, etc.).</li>
    <li><code>[Router]</code> for changes to the <code>vllm_router</code> (e.g., routing algorithm, router observability, etc.).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>Pass all linter checks. Please use <code>pre-commit</code> to format your code. See <code>README.md</code> for installation.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li> Please include sufficient tests to ensure the change is stay correct and robust. This includes both unit tests and integration tests.</li>
</ul>

<h3>DCO and Signed-off-by</h3>
<p>When contributing changes to this project, you must agree to the <a href="https://github.com/vllm-project/vllm/blob/main/DCO">DCO</a>. Commits must include a <code>Signed-off-by:</code> header which certifies agreement with the terms of the DCO.</p>
<p>Using <code>-s</code> with <code>git commit</code> will automatically add this header.</p>

<h3>What to Expect for the Reviews</h3>

### CODE REVIEW

**Comment by Xunzhuo** [comment]
@ppppqp the inference is running inside candle? Not by external calling?

**Comment by Xunzhuo** [comment]
For the follow up that you mentioned, I think we have already had the preference signal now in dashboard?

**Comment by ppppqp** [comment]
> @ppppqp the inference is running inside candle? Not by external calling?

Yep, inside candle. Specifically: `generative_classifier.classify_qwen3_preference`

**Comment by Xunzhuo** [comment]
And I wonder why not doing the SFT to better align with the classification task? And how is the latency that we run this in local vs in vLLM in large prompt and high batch size?

**Comment by ppppqp** [comment]
> For the follow up that you mentioned, I think we have already had the preference signal now in dashboard?

Sorry, i meant preference model configuration
Something like this:

image

**Comment by ppppqp** [comment]
> And I wonder why not doing the SFT to better align with the classification task? And how is the latency that we run this in local vs in vLLM in large prompt and high batch size?

Great questions

For the SFT part, I tried the local finetuned model and it fell short of several aspects:
1. The training data I used is multi-turn, but in reality we supply only the latest conversation turn to model
2. I trained on Qwen3-0.6B, and it looks like both fine-tuned version and base 0.6B version do not work that well. Qwen3-1.7B is a lot better than both models.
I did not fully aligned the training process with our inference methodology, so that's a potential reason that SLM does not work that well. I can continue exploring this path!

For the latency part, will follow up with some metrics!

File: candle-binding/src/model_architectures/generative/qwen3_multi_lora_classifier.rs [hunk 1]
```diff
@@ -1015,6 +1016,206 @@ impl Qwen3MultiLoRAClassifier {
...
+            });
+        }
+
+        // Clear KV cache before new classification
+        self.base_model.clear_kv_cache();
+
+        // Format prompt for zero-shot classification
+        let categories_str = categories.join(", ");
+        let instruction = format!(
+            "You are an expert classifier. Classify the following into exactly ONE category. Respond with ONLY the category name.\n\nCategories: {}\n\nClassify:\n{}\nAnswer:",
```
Comment (rootfs) [inline candle-binding/src/model_architectures/generative/qwen3_multi_lora_classifier.rs:1042]: is the output always in this format? what if the model outputs a different formt?

**Comment by ppppqp** [comment]
Thanks for reviewing!

There are a few changes I would like to make before merging this:
- I will add some golden output correctness check test cases
- I will add a benchmark for the implementation under `perf/benchmarks` and compare with our existing classifiers
- I will remove the preference classifier from the template config file, so that we can release it on v0.2, since we still want to explore the finetuned SLM path.

I'm also working on performance improvement on this implementation. Given that this PR is already reviewed, I will put up another PR on that!

**Comment by rootfs** [comment]
@ppppqp sounds good! followup will be great to catch them up.

**Comment by ppppqp** [comment]
@Xunzhuo I added some benchmark results on the PR description. Looks like on average latency is about 2~3 times of the classifier, which makes sense because 
1. This is a larger model (1.7B vs 0.6B)
2. Preference label can be multi-token

Do you think if it's reasonable? I will also try optimize the latency in the following PRs.

File: src/vllm-sr/cli/templates/router-defaults.yaml [hunk 2]
```diff
@@ -66,6 +66,13 @@ classifier:
     use_cpu: true
     pii_mapping_path: "models/mmbert32k-pii-detector-merged/pii_type_mapping.json"
 
+  # TODO: Enable preference classifier on V2
+  # preference_model:
+  #   model_id: "models/mom-preference-classifier"
```
Comment (Xunzhuo) [inline src/vllm-sr/cli/templates/router-defaults.yaml:71]: have you register the model into register.go if you want router to auto download the model with the name?

Comment (ppppqp) [reply src/vllm-sr/cli/templates/router-defaults.yaml:71]: Yea, I intentionally commented this out because I suppose we want to use a finetuned Qwen3 model. I also removed the preference model entry I added earlier in `registry.go`. Will add it back once we have a concensus on which model to use, but before that I don't want user to download the model to local, because we currently don't have a good way to cleaning local cache if we update the underlying model (but keeps using the same name as `models/mom-preference-classifier`

Do you think if it makes sense?

File: candle-binding/src/model_architectures/generative/qwen3_multi_lora_classifier.rs [hunk 3]
```diff
@@ -1015,6 +1016,206 @@ impl Qwen3MultiLoRAClassifier {
...
+                            logits_vec.len()
+                        ),
+                        input_context: Some(format!("category={}", categories[cat_idx])),
+                    });
+                }
+                let logit = logits_vec[token_id as usize];
+                // for each label token, compute log-probability
+                let probs = softmax(&logits_vec);
+                let prob = probs[token_id as usize].max(f32::MIN_POSITIVE);
+                total_log_prob += prob.ln();
```
[the remark this answers is not in the document]
Comment (ppppqp) [reply candle-binding/src/model_architectures/generative/qwen3_multi_lora_classifier.rs:1194]: Will try this in next PR!

File: dashboard/frontend/src/pages/ConfigPage.tsx [hunk 4]
```diff
@@ -1329,6 +1331,105 @@ const ConfigPage: React.FC<ConfigPageProps> = ({ activeSection = 'signals' }) =>
...
+                  use_cpu: false,
+                  use_qwen3: true,
+                },
+                [
+                  {
+                    name: 'model_id',
+                    label: 'Model ID or Path',
+                    type: 'text',
+                    required: true,
+                    placeholder: '',
```
[the remark this answers is not in the document]
Comment (ppppqp) [reply dashboard/frontend/src/pages/ConfigPage.tsx:1364]: Will improve in next PR!
